File: src/main/java/FastSlowPointer/Q4.java
package FastSlowPointer;
//Problem Challenge 1 - Palindrome LinkedList
/*
* The "Palindrome LinkedList" problem involves determining whether
* a given linked list is a palindrome or not. A palindrome is a sequence
* of characters or numbers that reads the same forward and backward.
* */
public class Q4 {

    public static class ListNode3 {
        int val;
        ListNode3 next;

        ListNode3(int x) {
            this.val = x;
            this.next = null;
        }
    }

    public static ListNode3 findMiddle(ListNode3 head) {
        ListNode3 slow = head;
        ListNode3 fast = head;

        while(fast != null && fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    public static ListNode3 reverseList(ListNode3 head) {
        ListNode3 prev = null;
        ListNode3 curr = head;

        while(curr != null) {
            ListNode3 next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }

    public static boolean isPalindrome(ListNode3 head) {
        if(head == null || head.next == null) {
            return true;
        }

        ListNode3 middle = findMiddle(head);
        ListNode3 secondHalfReversed = reverseList(middle.next);
        ListNode3 p1 = head;
        ListNode3 p2 = secondHalfReversed;

        while(p2 != null) {
            if(p1.val != p2.val) {
                return false;
            }

            p1 = p1.next;
            p2 = p2.next;
        }

        return true;
    }

    public static void main(String[] args) {

        ListNode3 head1 = new ListNode3(1);
        head1.next = new ListNode3(2);
        head1.next.next = new ListNode3(3);
        head1.next.next.next = new ListNode3(2);
        head1.next.next.next.next = new ListNode3(1);

        System.out.println(isPalindrome(head1));


    }
}

File: scala/FastSlowPointer/Q4.scala
class ListNode(var _x: Int = 0) {
    var next: ListNode = null
}
//Problem Challenge 2 - Rearrange a LinkedList (medium)
def reorderList(head: ListNode): Unit = {
    def reverseList(node: ListNode): ListNode = {
        var prev: ListNode = null
        var curr = node

        while(curr != null) {
            val nextTemp = curr.next
            curr.next = prev
            prev = curr
            curr = nextTemp
        }
        prev
    }

    def mergeLists(l1: ListNode, l2: ListNode): Unit = {
        var p1 = l1
        var p2 = l2

        while(p1 != null && p2 != null) {
            val nextTemp1 = p1.next
            val nextTemp2 = p2.next
            p1.next = p2
            p2.next = nextTemp1
            p1 = nextTemp1
            p2 = nextTemp2
        }
        
    }

    var slow = head
    var fast = head

    while(fast != null && fast.next != null) {
        slow = slow.next
        fast = fast.next.next
    }

    var secondHalf = reverseList(slow)
    mergeLists(head, secondHalf)
}

def main(args: Array[String]): Unit = {
    
    val head = new ListNode(1)
    val node1 = new ListNode(2)
    val node2 = new ListNode(3)
    val node3 = new ListNode(4)
    val node4 = new ListNode(5)

    head.next = node1
    node1.next = node2
    node2.next = node3
    node3.next = node4

    reorderList(head)

    println("Reordered List: ")
    var curr = head
    while(curr != null) {
        println(curr._x)
        curr = curr.next
    }
}
